test/simple/test-tls-cipher-list.js

```diff
@@ -60,11 +45,185 @@ function doTest(checklist, env, useswitch) {
+         ]);
+
+   // test the right-most command line option takes precedence
+   doTest('XYZ',
+          [
+            '--enable-legacy-cipher-list=v0.10.38',
+            '--cipher-list=XYZ'
+          ]);
+
+    // test the right-most command line option takes precedence
+    doTest('XYZ',
+           [
+             '--cipher-list=XYZ',
+             '--enable-legacy-cipher-list=v0.10.38',
+             '--cipher-list=XYZ'
+           ]);
+
+    // test that NODE_LEGACY_CIPHER_LIST takes precedence over
+    // NODE_CIPHER_LIST
+
+    doTest(V1038Ciphers, [],
+           {
+             'NODE_LEGACY_CIPHER_LIST': 'v0.10.38',
+             'NODE_CIPHER_LIST': 'ABC'
+           });
+
+}
+
+// Start running the tests...
+
 doTest(crypto.DEFAULT_CIPHER_LIST); // test the default
-doTest('ABC', {'NODE_CIPHER_LIST':'ABC'}); // test the envar
-doTest('ABC', 'ABC', 1); // test the --cipher-list switch
 
-['v0.10.38'].forEach(function(ver) {
-  doTest(tls.getLegacyCiphers(ver), ver, 2);
-  doTest(tls.getLegacyCiphers(ver), ver, 3);
+// Test the NODE_CIPHER_LIST environment variable
+doTest('ABC', [], {'NODE_CIPHER_LIST':'ABC'});
+
+// Test the --cipher-list command line switch
+doTest('ABC', ['--cipher-list=ABC']);
+
+// Test the --enable-legacy-cipher-list and NODE_LEGACY_CIPHER_LIST envar
+['v0.10.38'].forEach(function(arg) {
+  var checklist = tls.getLegacyCiphers(arg);
+  // command line switch
+  doTest(checklist, ['--enable-legacy-cipher-list=' + arg]);
+  // environment variable
+  doTest(checklist, [], {'NODE_LEGACY_CIPHER_LIST': arg});
+});
+
+// Test the precedence order for the various options
+doTestPrecedence();
+
+// Test that we throw properly
+// invalid value
+assert.throws(function() {tls.getLegacyCiphers('foo');}, Error);
+// no parameters
+assert.throws(function() {tls.getLegacyCiphers();}, TypeError);
+// not a string parameter
+assert.throws(function() {tls.getLegacyCiphers(1);}, TypeError);
+// too many parameters
+assert.throws(function() {tls.getLegacyCiphers('abc', 'extra');}, TypeError);
+// ah, just right
+assert.doesNotThrow(function() {tls.getLegacyCiphers('v0.10.38');});
+
+
+
+// Test to ensure default ciphers are not set when v0.10.38 legacy cipher
+// switch is used. This is a bit involved... we need to first set up the
+// TLS server, then spawn a second node instance using the v0.10.38 cipher,
+// then connect and check to make sure the options are correct. Since there
+// is no direct way of testing it, an alternate createCredentials shim is
+// created that intercepts the call to createCredentials and checks the output.
+// The following server code was adopted from test-tls-connect-simple.
+
+// note that the following function is written out to a string and
+// passed in as an argument to a child node instance.
+var script = (
+  function() {
+    var tls = require('tls');
+    var orig_createCredentials = require('crypto').createCredentials;
+    require('crypto').createCredentials = function(options) {
```

Comment:
That's not quite what this is testing. This tests to make sure the default
ciphers are not set on the options when v0.10.38 legacy is set. The monkey
patch allows us to intercept the call and verify that the ciphers property
is indeed not set.
On May 4, 2015 6:05 PM, "Julien Gilli" <notifications@github.com> wrote:

> In test/simple/test-tls-cipher-list.js
> <https://github.com/joyent/node/pull/14572#discussion_r29638490>:
>
> > +
> > +// Test to ensure default ciphers are not set when v0.10.38 legacy cipher
> > +// switch is used. This is a bit involved... we need to first set up the
> > +// TLS server, then spawn a second node instance using the v0.10.38 cipher,
> > +// then connect and check to make sure the options are correct. Since there
> > +// is no direct way of testing it, an alternate createCredentials shim is
> > +// created that intercepts the call to createCredentials and checks the output.
> > +// The following server code was adopted from test-tls-connect-simple.
> > +
> > +// note that the following function is written out to a string and
> > +// passed in as an argument to a child node instance.
> > +var script = (
> > +  function() {
> > +    var tls = require('tls');
> > +    var orig_createCredentials = require('crypto').createCredentials;
> > +    require('crypto').createCredentials = function(options) {
>
> Instead of monkey patching crypto.createCredentials, what do you think
> about starting a server that sets its ciphers list to RC4-MD5 only?
>
> Since it's explicitly excluded from the ciphers list after node v0.10.38,
> the test could make sure that the client connection can only be established
> when passing --enable-legacy-cipher-list=v0.10.38 or setting equivalent
> environment variables.
>
> I believe it would make the implementation of this test much simpler.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joyent/node/pull/14572/files#r29638490>.
>